include/DDMinimizer.hpp

```diff
@@ -0,0 +1,180 @@
+   * in the controlled gates.
+   * @param QuantumComputation
+   */
+  static void optimizeInputPermutation(qc::QuantumComputation& qc);
+
+  /**
+   * @brief Computes a permutation for the QuantumComputation based on a
+   * heuristic to optimize the running time of the DD-simulator.
+   * @param QuantumComputation
+   * @return the qc::Permutation The computed permutation to be used as the
+   * initialLayout for a QuantumComputation
+   * @details First collects operation indices of controlled operation for
+   * patterns (s. makeDataStructure). Then, based on the pattern of the
+   * controlled gates, the layout is adjusted. If no pattern is found, the
+   * control based permutation is created.
+   */
+  static qc::Permutation createGateBasedPermutation(qc::QuantumComputation& qc);
+
+  // Helper functions for createGateBasedPermutation
+  /**
+  * @brief creates a data structure for the pattern analysis of controlled gates
+  * @param QuantumComputation
+  * @return a pair of maps to save indices of controlled gates
+  * @details the data structure consists of two maps:
+  * 1. map: string of ladder (step) name to control and target bit to the index
+  of the operation
+  * 2. map: string of ladder name to vector: max index of operation for each
+  step or the c_x or x_c ladder
+  * for c_l and x_l position 0 in the vector marks the line of c(x) at 0 -> we
+  count the left most as the first
+  * for c_r and x_r position 0 in the vector marks the line of c(x) at bits - 1
+  -> we count theright most as the first
+  *
+  * The ladder (steps) describe the following controlled gates (c: control
+  qubit, x: target qubit): e.g. for three qubits
+        * c_x: c | 0  1  2
+               x | 1  2  3
+
+        * x_c: c | 1  2  3
+               x | 0  1  2
+
+        * c_l_1: c | 0  0  0  and  c_l_2: c | 1  1  and  c_l_3: c | 2
+                 x | 1  2  3              x | 2  3              x | 3
+
+        * c_r_1: c | 3  3  3  and  c_r_2: c | 2  2  and  c_r_3: c | 1
+                 x | 0  1  2              x | 0  1              x | 0
+
+        * x_l_1: c | 1  2  3  and  x_l_2: c | 2  3  and  x_l_3: c | 3
+                 x | 0  0  0              x | 1  1              x | 2
+
+        * x_r_1: c | 0  1  2  and  x_r_2: c | 0  1  and  x_r_3: c | 0
+                 x | 3  3  3              x | 2  2              x | 1
+
+  */
+  static std::pair<
+      std::map<std::string, std::map<std::pair<Qubit, Qubit>, int>>,
+      std::map<std::string, std::vector<int>>>
+  makeDataStructure(qc::QuantumComputation& qc);
+
+  static void initializeDataStructure(
+      std::size_t bits, std::map<std::pair<Qubit, Qubit>, int>& xCMap,
+      std::map<std::pair<Qubit, Qubit>, int>& cXMap,
+      std::vector<std::map<std::pair<Qubit, Qubit>, int>>& cLMap,
+      std::vector<std::map<std::pair<Qubit, Qubit>, int>>& cHMap,
+      std::vector<std::map<std::pair<Qubit, Qubit>, int>>& xLMap,
+      std::vector<std::map<std::pair<Qubit, Qubit>, int>>& xHMap);
```

Comment:
When I am looking at this function with very special arguments, I wonder whether it might be better to have the variables that are arguments right now as members of the class. Then, the function would not have any arguments but, when called, would still initialize the class members.

Additionally: Does that function need to be public, or can it be private? This already applies to the function above, I guess.